ghproxy/ghcache/ghcache.go

```diff
@@ -138,18 +139,122 @@ func cacheResponseMode(headers http.Header) CacheResponseMode {
+}
+
+// tokenRegistry keeps the timestamp of last handled request per token budget (appId or hash)
+type tokensRegistry struct {
+	lock              sync.Mutex
+	tokens            map[string]tokenInfo
+	throttlingTime    time.Duration
+	getThrottlingTime time.Duration
+}
+
+func (tr *tokensRegistry) getRequestWaitDuration(tokenBudgetName string, getReq bool) time.Duration {
+	var duration time.Duration
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+	toQueue := time.Now()
+	if t, exists := tr.tokens[tokenBudgetName]; exists {
+		toQueue, duration = tr.calculateRequestWaitDuration(t, toQueue, getReq)
+	}
+	tr.tokens[tokenBudgetName] = tokenInfo{getReq: getReq, timestamp: toQueue}
+	return duration
+}
+
+func (tr *tokensRegistry) calculateRequestWaitDuration(t tokenInfo, toQueue time.Time, getReq bool) (time.Time, time.Duration) {
+	throttlingTime := tr.throttlingTime
+	followingGetRequests := t.getReq && getReq
+	// Previous request also was GET => use GET throttling time
+	if followingGetRequests {
+		throttlingTime = tr.getThrottlingTime
+	}
+	duration := toQueue.Sub(t.timestamp)
+
+	if toQueue.Before(t.timestamp) || toQueue.Equal(t.timestamp) {
+		// There is already queued request, queue next afterwards.
+		difference := throttlingTime
+		if getReq && !followingGetRequests {
+			difference = tr.getThrottlingTime
+		}
+		future := t.timestamp.Add(difference)
+		duration = future.Sub(toQueue)
+		toQueue = future
+	} else if duration >= throttlingTime {
+		// There was no request for some time, no need to wait.
+		duration = 0
+	} else {
+		// There is a queued request, wait until the next throttling tick.
+		duration = throttlingTime - duration
+		if getReq && !followingGetRequests {
+			duration = tr.getThrottlingTime
+		}
+		toQueue = toQueue.Add(duration)
+	}
+	return toQueue, duration
 }
 
-// throttlingTransport throttles outbound concurrency from the proxy
+// throttlingTransport throttles outbound concurrency from the proxy and adds QPS limit (1 request per given time) if enabled
 type throttlingTransport struct {
-	sem          *semaphore.Weighted
-	roundTripper http.RoundTripper
+	sem                      *semaphore.Weighted
+	roundTripper             http.RoundTripper
+	hasher                   ghmetrics.Hasher
+	requestThrottlingTime    uint
+	getRequestThrottlingTime uint
+	apiv3                    tokensRegistry
+	apiv4                    tokensRegistry
+}
+
+func (c *throttlingTransport) holdRequest(req *http.Request) {
+	var tokenBudgetName string
+	if val := req.Header.Get(TokenBudgetIdentifierHeader); val != "" {
+		tokenBudgetName = val
+	} else {
+		tokenBudgetName = c.hasher.Hash(req)
+	}
+	getReq := req.Method == http.MethodGet
+	var duration time.Duration
+	if strings.HasPrefix(req.URL.Path, "graphql") || strings.HasPrefix(req.URL.Path, "/graphql") {
+		duration = c.apiv4.getRequestWaitDuration(tokenBudgetName, getReq)
+	} else {
+		duration = c.apiv3.getRequestWaitDuration(tokenBudgetName, getReq)
+	}
+
+	if duration > 0 {
```

Comment:
I think `duration` would make a good metric to record somehow? If this grows too much then that is probably something the operator should be aware of.

Reply:
Added metrics per org (token budget name), API and request type (`GET`, `POST`, etc)